File: dom/nfc/NdefRecord.cpp
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "NdefRecord.h"
#include "nsIDOMClassInfo.h"

#if defined(MOZ_WIDGET_GONK)
#include <android/log.h>
#define LOG(args...)  __android_log_print(ANDROID_LOG_INFO, "Gonk", args)
#else
#define LOG(args...) 
#endif

DOMCI_DATA(MozNdefRecord, mozilla::dom::nfc::NdefRecord)

namespace mozilla {
namespace dom {
namespace nfc {

// For Query interface to find the mapped interface.
NS_INTERFACE_MAP_BEGIN(NdefRecord)
  NS_INTERFACE_MAP_ENTRY(nsIDOMMozNdefRecord)
  NS_INTERFACE_MAP_ENTRY(nsISupports) 
  NS_DOM_INTERFACE_MAP_ENTRY_CLASSINFO(MozNdefRecord)
NS_INTERFACE_MAP_END

NS_IMPL_ADDREF(NdefRecord)
NS_IMPL_RELEASE(NdefRecord)

/********************************
 * Constructors/Destructors
 ********************************/

NdefRecord::NdefRecord()
{
  LOG("DOM NdefRecord: No parameter instantiation");
  tnf = 0;
  payload = JSVAL_NULL;
}

nsresult NdefRecord::NewNdefRecord(nsISupports* *aNewObject)
{
  LOG("DOM nsNdefRecord: No parameter instantiation");
  NS_ADDREF(*aNewObject = new NdefRecord());
  return NS_OK;
}

/************************************
 * NS_METHODIMPs (Getter and Setter)
 ************************************/
NS_IMETHODIMP
NdefRecord::GetTnf(char* aTnf)
{
  LOG("DOM NdefRecord.GetTnf");
  *aTnf = tnf;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::SetTnf(const char aTnf)
{
  LOG("DOM NdefRecord.SetTnf");
  tnf = aTnf;
  return NS_OK; 
}

NS_IMETHODIMP
NdefRecord::GetType(nsAString& aType)
{
  LOG("DOM NdefRecord.GetType");
  aType = type;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::SetType(const nsAString & aType)
{
  LOG("DOM NdefRecord.SetType");
  type = aType;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::GetId(nsAString& aId)
{
  LOG("DOM NdefRecord.GetId");
  aId = id;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::SetId(const nsAString& aId)
{
  LOG("DOM NdefRecord.SetId");
  id = aId;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::GetPayload(JS::Value* aPayload)
{
  LOG("DOM NdefRecord.GetPayload");
  *aPayload = payload;
  return NS_OK;
}

NS_IMETHODIMP
NdefRecord::SetPayload(const JS::Value& aPlayload)
{
  LOG("DOM NdefRecord.SetPayload");
  payload = aPlayload;
  return NS_OK;
}

} // namespace nfc
} // namespace dom
} // namespace mozilla

File: dom/nfc/NDEFRecord.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

/* Copyright © 2014, Deutsche Telekom, Inc. */

"use strict";

const DEBUG = true;
function debug(s) {
  if (DEBUG) dump("-*- Nfc DOM: " + s + "\n");
}

const Cc = Components.classes;
const Ci = Components.interfaces;
const Cu = Components.utils;

Cu.import("resource://gre/modules/XPCOMUtils.jsm");

/**
 * MozNDEFRecord
 */
function MozNDEFRecord() {
  debug("In MozNDEFRecord Constructor");
}
MozNDEFRecord.prototype = {
  _window: null,

  _tnf: 0x0,
  _type: null,
  _id: null,
  _payload: null,

  // nsIDOMGlobalPropertyInitializer
  init: function nr_init(aWindow) {
    debug("MozNDEFRecord: init()");
    this._window = aWindow;
  },

  __init: function nr___init(aTnf, aType, aId, aPayload) {
   debug("MozNDEFRecord: __init()");
   this._tnf = aTnf;
   this._type = aType;
   this._id = aId;
   this._payload = aPayload;
  },

  // nsISystemMessagesWrapper...but this is not a message.
  wrapMessage: function nr_wrapMessage(aMessage, aWindow) {
    //return Cu.cloneInto(this, this._window);
  }, 

  initialize: function nr_initialize(aWindow, tnf, type, id, payload) {
    debug("MozNDEFRecord: initialize()"); 
    this._window = aWindow;

    this.tnf = tnf;
    this.type = type;
    this.id = id;
    this.payload = payload;
  },

  get tnf() {
    return this._tnf;
  },

  get type() {
    return this._type;
  },

  get id() {
    return this._id;
  },

  get payload() {
    return this._payload;
  },
  
  doSomething: function nr_doSomething() {
    debug("doSomething");
  },

  classID: Components.ID("{b0687a80-11fe-11e4-9191-0800200c9a66}"),
  contractID: "@mozilla.org/nfc/NDEFRecord;1",
  QueryInterface: XPCOMUtils.generateQI([Ci.nsISupports,
                                         Ci.nsIDOMGlobalPropertyInitializer,
                                         Ci.nsISystemMessagesWrapper]),
};

this.NSGetFactory = XPCOMUtils.generateNSGetFactory([MozNDEFRecord]);
